### PR TITLE
Implementation for non-collinear and spin-orbit calculations in vasp.6.* 

### PR DESCRIPTION
**Pull request by dariofiosca**
This pull request is to adapt triqs for noncollinear calculations done with vasp.6.1 and later versions. It can handle both noncollinear (without soc) and with soc effect. Tutorials regarding the utilization will soon come.

It has not yet been tested on full charge self-consistent calculations

An extra test case for Ce compound is in the test folder.

### CODE REVIEW

File: python/triqs_dft_tools/converters/plovasp/elstruct.py [hunk 1]
```diff
@@ -61,6 +61,8 @@ def __init__(self, vasp_data):
 
         self.kmesh = {'nktot': self.nktot}
         self.kmesh['kpoints'] = vasp_data.kpoints.kpts
+        # VASP.6.
+        self.nc_flag = vasp_data.plocar.nc_flag
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/elstruct.py:65]: this only works for Vasp 6 right?

File: python/triqs_dft_tools/converters/plovasp/elstruct.py [hunk 2]
```diff
@@ -80,8 +82,7 @@ def __init__(self, vasp_data):
 # Note that the number of spin-components of projectors might be different from those
 # of bands in case of non-collinear calculations
         self.nspin = vasp_data.plocar.nspin
-        self.nc_flag = vasp_data.plocar.ncdij == 4
-
+        #self.nc_flag = vasp_data.plocar.ncdij == 2
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/elstruct.py]: then we can remove this line.

File: python/triqs_dft_tools/converters/plovasp/proj_group.py [hunk 3]
```diff
@@ -416,8 +416,9 @@ def orthogonalize_projector_matrix(self, p_matrix):
         overlap = np.dot(p_matrix, p_matrix.conj().T)
 # Calculate [O^{-1/2}]_{m m'}
         eig, eigv = np.linalg.eigh(overlap)
-        assert np.all(eig > 0.0), ("Negative eigenvalues of the overlap matrix:"
-           "projectors are ill-defined")
+        eig = np.around(eig,10)  
+        #assert np.all(eig > 0.0), ("Negative eigenvalues of the overlap matrix:"
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/proj_group.py:420]: I guess for NC calculations this check is not useful, but we should keep it for normal calculations I guess?

File: python/triqs_dft_tools/converters/plovasp/proj_shell.py [hunk 4]
```diff
@@ -252,18 +242,13 @@ def extract_projectors(self, proj_raw, proj_params, kmesh, structure):
             self.proj_arr = np.zeros((nion, ns, nk, nlm, nb), dtype=np.complex128)
             for io, ion in enumerate(self.ion_list):
                 qcoord = structure['qcoords'][ion]
-                for m in range(nlm):
+                print(nlm)
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/proj_shell.py]: remove debug print?

File: python/triqs_dft_tools/converters/plovasp/proj_shell.py [hunk 5]
```diff
@@ -322,15 +316,26 @@ def density_matrix(self, el_struct, site_diag=True, spin_diag=True):
...
+                        array_sum = np.sum(proj_k)
+                        if np.isnan(array_sum) == True: 
+                            count_nan += 1
+                        if self.nc_flag == True:
+                            occ_mats[isp, io, :, :] += 0.5 * np.dot(proj_k * occ[ib1:ib2], proj_k.T.conj()).real * weight
+                            overlaps[isp, io, :, :] += np.dot(proj_k, proj_k.T.conj()).real * weight
+                        else:
+                            occ_mats[isp, io, :, :] += np.dot(proj_k * occ[ib1:ib2], proj_k.T.conj()).real * weight
+                            overlaps[isp, io, :, :] += np.dot(proj_k, proj_k.T.conj()).real * weight
+            assert count_nan == 0, "!!! WARNING !!!: There are %s  NaN in your Projectors"%(count_nan)
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/proj_shell.py:336]: Ah I guess this is now the relevant check? Do we also need to check if there are negative / imaginary entries? If so, please remove the commented lines above completely.

File: python/triqs_dft_tools/converters/plovasp/proj_shell.py [hunk 6]
```diff
@@ -339,13 +344,11 @@ def density_matrix(self, el_struct, site_diag=True, spin_diag=True):
                         i1 = io * nlm
                         i2 = (io + 1) * nlm
                         proj_k[i1:i2, :] = self.proj_win[io, isp, ik, ...]
+                    # DARIO TODO
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/proj_shell.py]: ?

File: python/triqs_dft_tools/converters/plovasp/vaspio.py [hunk 7]
```diff
@@ -140,84 +140,6 @@ def from_file(self, vasp_dir='./', plocar_filename='PLOCAR'):
 #        self.proj_params, self.plo = self.temp_parser(projcar_filename=vasp_dir + "PROJCAR", locproj_filename=vasp_dir + "LOCPROJ")
         self.proj_params, self.plo = self.locproj_parser(locproj_filename=vasp_dir + "LOCPROJ")
 
-    def temp_parser(self, projcar_filename='PROJCAR', locproj_filename='LOCPROJ'):
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/vaspio.py:143]: I guess this is not used anyway right?

File: python/triqs_dft_tools/converters/plovasp/vaspio.py [hunk 8]
```diff
@@ -242,8 +164,13 @@ def lm_to_l_m(lm):
             line = f.readline()
             line = line.split("#")[0]
             sline = line.split()
-            self.ncdij, nk, self.nband, nproj = list(map(int, sline[:4]))
-            self.nspin = 1 if self.ncdij == 1 else 2
+            print(sline)
```
Comment (the-hampel) [inline python/triqs_dft_tools/converters/plovasp/vaspio.py]: debug print?

File: test/python/plovasp/converter/test_ce_noncollinear.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    def generate_and_output_as_text(pars, el_struct)
+        and
+    VaspConverter
+
+    Scenarios:
+
+    - Parse config file and produce a correct converted h5-file
+    """
+# Scenario 1
```
Comment (the-hampel) [inline test/python/plovasp/converter/test_ce_noncollinear.py:27]: I believe we do not need all the files for the test to work right? For example the vasp.ctrl and vasp.pg1 file will be generated. Can we remove them from the test files?

**Review by the-hampel — COMMENTED** [review]
Hi @dariofiosca , 
thanks for the PR. That looks really good. I made some general comments that you maybe can very quickly address. I also removed the .ipynb checkpoint files and the bash install script. We do not need them here. :-) While you have a look at the comments I will try to run the new converter. Thank you!

File: python/triqs_dft_tools/converters/plovasp/vaspio.py (hunk 8 above)
Comment (dariofiosca) [reply python/triqs_dft_tools/converters/plovasp/vaspio.py]: yes

File: test/python/plovasp/converter/test_ce_noncollinear.py (hunk 9 above)
Comment (dariofiosca) [reply test/python/plovasp/converter/test_ce_noncollinear.py:27]: They can be removed

File: python/triqs_dft_tools/converters/plovasp/elstruct.py (hunk 1 above)
Comment (dariofiosca) [reply python/triqs_dft_tools/converters/plovasp/elstruct.py:65]: It works only for VASP.6.1 or later versions

File: python/triqs_dft_tools/converters/plovasp/elstruct.py (hunk 2 above)
Comment (dariofiosca) [reply python/triqs_dft_tools/converters/plovasp/elstruct.py]: The commented one ? Yes

File: python/triqs_dft_tools/converters/plovasp/proj_shell.py (hunk 4 above)
Comment (dariofiosca) [reply python/triqs_dft_tools/converters/plovasp/proj_shell.py]: Yes

**Pull request closed without merging**